keps/sig-node/4800-cpumanager-split-uncorecache/README.md

```diff
@@ -0,0 +1,686 @@
+
+## Motivation
+
+Our motivation is to reduce noisy neighbor scenarios that occur on systems with split uncore cache, which is available on both x86 and ARM architecture.
+The challenge with current kubelet’s cpu manager is that it is unaware of split uncore architecture and spreads CPU assignments across distributed uncore caches. This creates a noisy neighbor problem where multiple pods/containers are sharing the same uncore cache. In addition, pods spread against multiple uncore caches sees higher latency and reduced performance due to inter-cache access latency. For workload use cases that are sensitive to latency and are performance deterministic, minimizing the noisy neighbor condition in uncore cache can have significant improvements in performance.
+
+![Figure to the motivation](https://github.com/ajcaldelas/enhancements/blob/split_l3_cache/keps/sig-node/4800-cpumanager-split-uncorecache/images/uncore-alignment-motivation.png)
+
+The figure below highlights performance gain when pod placement is aligned to an uncore cache group. In this example HammerDB TPROC-C/My-SQL is the workload deployed. There was a 18% uplift in performance when uncore cache aligned compared to default behavior. Other workloads may see higher gains.
+
+![Figure to the results of uncore cache alignment](https://github.com/ajcaldelas/enhancements/blob/split_l3_cache/keps/sig-node/4800-cpumanager-split-uncorecache/images/results.PNG) ![Figure of results](https://github.com/ajcaldelas/enhancements/blob/c7e423439608001cc837e2f5629dc87935353611/keps/sig-node/4800-cpumanager-split-l3cache/images/chart.png)
+
+### Goals
+
+- Introduce a new CPU Manager policy option that assigns CPU within the same grouping of uncore cache to container scope.  
+- Minimize the number of CPUs assignments from different uncore cache grouping to pods and container scope.
+
+### Non-Goals
+
+- This proposal does not aim to modify CPU assignments for CPU Manager policy set to none.
+- This does not alter the behavior of existing static policy options such as full-pcpus-only.
+
+## Proposal
+
+- Add a new static policy option to fine-tune of CPU assignments to align by uncore cache grouping.
+
+- Topology struct already contains CPUDetails which is a map of CPUInfo.  CPUInfo knows about NUMA, socket, and core IDs associated with a CPU.  We just need to add a new member called uncorecacheId into the CPUInfo struct that tracks whether the CPU is part of a split uncore caches. Functionality to support this was merged in cadvisor with these pull-requests [cadvisor/pr-2849](https://github.com/google/cadvisor/pull/2849) and [cadvisor/pr-2847](https://github.com/google/cadvisor/pull/2847/)
+
+- Handle enablement of the policy option in pkg/kubelet/cm/cpumanager/policy_options.go and check the validity of the user flag and capability of the system.  If topology does not support split uncore caches, grouping by uncore cache static policy will be ignored.  
```

Comment:
Do we warn when we are unable to align? Should the name be prefixed with `prefer` so in the future we may consider hard failing when conditions aren't met?

Reply:
The implementation is preferred so in the case where the node is highly utilized and the available CPUs are fractured across multiple uncore caches, the container will still be scheduled as long as the CPU limit can be met. We can add the prefix to make that more clear.

Reply:
I'd be slightly in favor of renaming `prefer-align-cpus-by-uncorecache` or any shorter name which conveys the intent.